Forms/FormBuilder.php

```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace SilverStripe\Forms;
+
+use SilverStripe\Control\Controller;
+use SilverStripe\ORM\DataObject;
+
+/**
+ * A service which can generate a form for a given record and controller
+ */
+interface FormBuilder {
+
+	/**
+	 * @return DataObject
+	 */
+	public function getRecord();
+
+	/**
+	 * @return Controller
+	 */
+	public function getController();
```

Comment:
I think the problem here is that we are lacking a way of providing context for a formbuilder. In extensions it will be necessary not to simply know that a form is being scaffolded, but also why and how. For instance, we will need a way to distinguish between a userform scaffolding a front-end form and a cms-edit form. How will extensions on UserForm know how to determine this context and update the form accordingly? In each case, the same record is being provided, but a different Controller (CMSEditPageController vs UserForm_Controller). Thus it would need to query the getController() method on the second argument to the `updateFormFields` method.

Likewise, the context of the record itself is necessary; Certain records will have different fields, and thus extensions on the controller may need to respond. For instance, an extension on ModelAdmin may need to know if the underlying record has versioned or not (and thus provide additional publishing controls), thus it would query the getRecord() method for this use case.